From fa3beda9fcaffdbd3afb7049c8013f18b24f524b Mon Sep 17 00:00:00 2001
From: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Date: Thu, 15 Feb 2024 19:13:30 +0100
Subject: [PATCH 1013/1695] pta: stm32mp: remoteproc: implement the
 PTA_REMOTEPROC_GET/RELEASE_MEM API

The PTA_REMOTEPROC_GET_MEM is used to get access right to the remote
processor memories. This should be called before accessing to the
memories for loading or cleaning.

The PTA_REMOTEPROC_RELEASE_MEM is used to release the access rights.
The memory are re-assigned to the remote processor.

These services are mandatory and the PTA should implements void
functions if no supported.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 544bb96086c624e96ede8241da00150d52ed517d pta: stm32mp: remoteproc: implement the PTA_REMOTEPROC_GET/RELEASE_MEM API]
Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Change-Id: Ie333a95ad4690419648f21434d6e649540284279
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/361063
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/391545
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Tested-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
---
 core/pta/stm32mp/remoteproc_pta.c    | 38 ++++++++++++++++++++++++++++
 lib/libutee/include/remoteproc_pta.h | 18 +++++++++++++
 2 files changed, 56 insertions(+)

diff --git a/core/pta/stm32mp/remoteproc_pta.c b/core/pta/stm32mp/remoteproc_pta.c
index 5a124f550..47f74388c 100644
--- a/core/pta/stm32mp/remoteproc_pta.c
+++ b/core/pta/stm32mp/remoteproc_pta.c
@@ -355,6 +355,40 @@ static TEE_Result rproc_pta_clean(uint32_t pt, TEE_Param params[TEE_NUM_PARAMS])
 	return stm32_rproc_clean(params[0].value.a);
 }
 
+static TEE_Result rproc_pta_get_mem(uint32_t pt,
+				    TEE_Param params[TEE_NUM_PARAMS])
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_NONE);
+
+	if (pt != exp_pt)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (rproc_ta_state != REMOTEPROC_OFF)
+		return TEE_ERROR_BAD_STATE;
+
+	return stm32_rproc_get_mem(params[0].value.a);
+}
+
+static TEE_Result rproc_pta_release_mem(uint32_t pt,
+					TEE_Param params[TEE_NUM_PARAMS])
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_NONE);
+
+	if (pt != exp_pt)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (rproc_ta_state != REMOTEPROC_OFF)
+		return TEE_ERROR_BAD_STATE;
+
+	return stm32_rproc_release_mem(params[0].value.a);
+}
+
 static TEE_Result rproc_pta_invoke_command(void *session __unused,
 					   uint32_t cmd_id,
 					   uint32_t param_types,
@@ -379,6 +413,10 @@ static TEE_Result rproc_pta_invoke_command(void *session __unused,
 		return rproc_pta_tlv_param(param_types, params);
 	case PTA_REMOTEPROC_CLEAN:
 		return rproc_pta_clean(param_types, params);
+	case PTA_REMOTEPROC_GET_MEM:
+		return rproc_pta_get_mem(param_types, params);
+	case PTA_REMOTEPROC_RELEASE_MEM:
+		return rproc_pta_release_mem(param_types, params);
 	default:
 		return TEE_ERROR_NOT_IMPLEMENTED;
 	}
diff --git a/lib/libutee/include/remoteproc_pta.h b/lib/libutee/include/remoteproc_pta.h
index 7868ae08d..85ea4122c 100644
--- a/lib/libutee/include/remoteproc_pta.h
+++ b/lib/libutee/include/remoteproc_pta.h
@@ -177,4 +177,22 @@ static inline size_t rproc_pta_keyinfo_size(struct rproc_pta_key_info *keyinf)
  */
 #define PTA_REMOTEPROC_CLEAN		10
 
+/*
+ * Get remote processor memories access right.
+ *
+ * Request the access rights to the remote processor memories.
+ *
+ * [in]  params[0].value.a:	Unique 32bit remote processor identifier
+ */
+#define PTA_REMOTEPROC_GET_MEM		11
+
+/*
+ * Release remote processor memories access right.
+ *
+ * Release the access rights to the remote processor memories.
+ *
+ * [in]  params[0].value.a:	Unique 32bit remote processor identifier
+ */
+#define PTA_REMOTEPROC_RELEASE_MEM	12
+
 #endif /* __REMOTEPROC_PTA_H */
-- 
2.47.2

